[Chan Zun Mun Terence] iP

Hackin7 opened the pull request.

Hackin7 committed:
Rename, add greet and exit

Hackin7 committed:
Add echo, bye

Hackin7 committed:
Add listing and adding to list functionality

Hackin7 committed:
Add and Switch to Task class

Hackin7 committed:
Add mark as done functionality

Hackin7 committed:
Rename variables to match code style

Hackin7 committed:
Improve code naming and structure

- Rename task list from `items` to `tasks`
- Move print task from Cuboyd to Task

Hackin7 committed:
Add new types of tasks

Hackin7 committed:
Refactor and Add Command Parser with argument support

Hackin7 committed:
Add support for tracking ToDos, Events, Deadlines

Hackin7 committed:
Fix test scripts and Add a test

Hackin7 committed:
Refactor common UI elements

Hackin7 committed:
Add TaskTrackerUI Singleton and Refactor

Hackin7 committed:
Refactor Command Parser

Hackin7 committed:
Remove redundant variables

yijiano commented on src/main/java/Cuboyd.java:
```diff
@@ -0,0 +1,58 @@
+public class Cuboyd {
+    // Helpers /////////////////////////////////////////////////////////////////////////////////////////////////////////
+    public static void displayIntroText(){
+        String name = "Cuboyd";
```
```suggestion
        String BOT_NAME = "Cuboyd";
```
Should the name of the bot be treated as a constant instead (i.e. be in uppercase)?

yijiano commented on src/main/java/Cuboyd.java:
```diff
@@ -0,0 +1,58 @@
+            line = scanner.nextLine();
+            argumentsList = CommandParser.parseCommandToArguments(line);
+
+            switch(argumentsList.get(CommandParser.ARGUMENT_COMMAND)){
+                case "list":
+                    taskTrackerUI.listTasks();
+                    break;
+                case "todo":
+                    taskTrackerUI.addTodo(argumentsList);
+                    break;
+                case "deadline":
+                    taskTrackerUI.addDeadline(argumentsList);
+                    break;
+                case "event":
+                    taskTrackerUI.addEvent(argumentsList);
+                    break;
+                case "mark":
+                    taskTrackerUI.markTask(argumentsList);
+                    break;
+                case "unmark":
+                    taskTrackerUI.unmarkTask(argumentsList);
+                    break;
+                case "bye":
+                    System.out.println("Bye. Hope to see you again soon!");
+                    isAskingInput = false;
+                    break;
+                default:
+                    System.out.println("No valid command given!");
+                    break;
+            }
```
Should the indentation for `case` clauses be removed?

yijiano commented on src/main/java/ToDo.java:
```diff
@@ -0,0 +1,8 @@
+    public ToDo(String description) {
+        super(description);
+    }
+    public String toString() {
```
I like how you overload the toString() method here.

yijiano commented on src/main/java/CommandParser.java:
```diff
@@ -0,0 +1,43 @@
+    public static final String ARGUMENT_COMMAND = "command";
+    public static final String ARGUMENT_MAIN = "main";
+    public static HashMap<String, String> parseCommandToArguments(String line) {
+        HashMap<String, String> argumentsList = new HashMap<>();
+        String[] lineArgs = line.split(" ");
+
+        // Command
+        if (lineArgs.length <= 0) {
+            argumentsList.put(CommandParser.ARGUMENT_COMMAND,"");
+            return argumentsList;
+        }
+        argumentsList.put(CommandParser.ARGUMENT_COMMAND,lineArgs[0]);
+
+        // Arguments
+        String currArgumentName = CommandParser.ARGUMENT_MAIN;
+        StringBuilder currArgument = new StringBuilder();
+
+        for (int i=1; i<lineArgs.length; i++) {
+            if (lineArgs[i].isEmpty()) { // Should be redundant but just in case
+                continue;
+            }
+            if (lineArgs[i].charAt(0) == '/') {
+                // New argument
+                if (!currArgument.toString().isEmpty()){
+                    argumentsList.put(currArgumentName, currArgument.toString().strip());
+                }
+                currArgumentName = lineArgs[i];
+                currArgument.setLength(0);
+            } else {
+                // Add on to existing argument
+                currArgument.append(" ").append(lineArgs[i]);
+            }
+        }
+        // Add last command
+        if (!currArgument.toString().isEmpty()) {
+            argumentsList.put(currArgumentName, currArgument.toString().strip());
+        }
+        return argumentsList;
+    }
+}
```
I like how your implementation for this class, as it would allow more commands to be added in the future.

yijiano commented on src/main/java/TaskTrackerUI.java:
```diff
@@ -0,0 +1,100 @@
+    public void listTasks(){
+        System.out.println("Here are the tasks in your list:");
+        for (int currentItemIndex = 0; currentItemIndex< this.tasks.size(); currentItemIndex++){
+            System.out.println(String.valueOf(currentItemIndex+1) + "." + this.tasks.get(currentItemIndex));
```
Should this print statement be split into multiple lines (wrapped lines)?

yijiano reviewed (commented):
LGTM. Just a few nits to fix with regards to code style. Keep up the good work!

Hackin7 committed:
Add clearer error handling

- Unindent case statements for commands
- Add command feedback
- Fixed a bug with the out of range check in parseValidTaskIndex
- Change `return index` to `return INVALID_TASK_INDEX` for clarity

Hackin7 committed:
Merge branch 'branch-Level-5'

Hackin7 committed:
Rename constant to meet standards

Hackin7 committed:
Change Error Handling to Exceptions

Hackin7 committed:
Add delete task functionality

Hackin7 committed:
Add load and save functionality

Hackin7 committed:
Clean up collections usage

Hackin7 committed:
Merge branch 'branch-Level-6'

Hackin7 committed:
Merge branch 'branch-Level-7'

Hackin7 committed:
Add autosaving to delete

Hackin7 committed:
Add Manifest

okkhoy commented on src/main/java/CommandParser.java:
```diff
@@ -0,0 +1,43 @@
+    public static final String ARGUMENT_MAIN = "main";
+    public static HashMap<String, String> parseCommandToArguments(String line) {
+        HashMap<String, String> argumentsList = new HashMap<>();
+        String[] lineArgs = line.split(" ");
+
+        // Command
+        if (lineArgs.length <= 0) {
+            argumentsList.put(CommandParser.ARGUMENT_COMMAND,"");
+            return argumentsList;
+        }
+        argumentsList.put(CommandParser.ARGUMENT_COMMAND,lineArgs[0]);
```
can SLAP this better.

okkhoy commented on src/main/java/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,3 @@
+Manifest-Version: 1.0
```
probably mistakenly moved inside the src/main/java folder

okkhoy commented on src/main/java/TaskTrackerUI.java:
```diff
@@ -0,0 +1,211 @@
+            System.out.println(e.getMessage());
+        }
+    }
+    // Helpers /////////////////////////////////////////////////////////////////////////////////////////////////////////
```
comments like these are redundant.

okkhoy commented on src/main/java/TaskTrackerUI.java:
```diff
@@ -0,0 +1,211 @@
+        System.out.println("Got it. I've added this task:");
+        System.out.println("  " + task.toString());
+        System.out.printf("Now you have %d tasks in the list.\n", tasks.size());
+    }
+    public int parseValidTaskIndex(HashMap<String, String> argumentsList, ArrayList<Task> tasks)
+    throws CuboydException{
```
could use some vertical spacing for better readability

okkhoy commented on src/main/java/TaskTrackerUI.java:
```diff
@@ -0,0 +1,211 @@
+    }
+    // Save & Load /////////////////////////////////////////////////////////////////////////////////////////////////////
+    public static final String SAVE_FILE_PATH = "./savedata.txt";
+    public void save() throws CuboydException {
```
looks like a lot of arrow-head code

okkhoy commented on src/main/java/TaskTrackerUI.java:
```diff
@@ -0,0 +1,211 @@
+                break;
+            case "D":
+                currentTask = new Deadline(
+                        new String(Base64.getDecoder().decode(list[2]), StandardCharsets.UTF_8),
+                        new String(Base64.getDecoder().decode(list[3]), StandardCharsets.UTF_8)
```
low level parsing can be extracted out as a method for better SLAP

okkhoy reviewed (commented):
Try to improve SLAP in the code

Hackin7 committed:
Extract Ui, Storage, Parser, TaskList classes

Hackin7 committed:
Add find functionality

Hackin7 committed:
Add JavaDoc

Hackin7 committed:
Merge pull request #1 from Hackin7/branch-Level-9

Add find functionality

Hackin7 committed:
Merge branch 'master' of https://github.com/Hackin7/ip into branch-A-JavaDoc

Hackin7 committed:
Update JavaDocs for listing and finding

Hackin7 committed:
Merge pull request #2 from Hackin7/branch-A-JavaDoc

Add JavaDoc

Hackin7 committed:
Update User guide

Hackin7 committed:
Remove help message for unimplemented commands

Hackin7 committed:
Refactor Storage to better follow SLAP

Hackin7 committed:
Remove redundant imports

Hackin7 committed:
Move MANIFEST.MF

Hackin7 committed:
Refactor Classes and Added more SLAP

Hackin7 committed:
Fix Compilation Issues in Tests

Hackin7 committed:
Fix extra printed line on savedata load failure

Hackin7 committed:
Fixed Test 1

Hackin7 committed:
Added Error Handling Test

Hackin7 committed:
Add Saving Tests

Hackin7 committed:
Add unknown command count in Error Handling Test

Hackin7 committed:
Improve variable names

Hackin7 committed:
Cleanup and Improve imports

Hackin7 committed:
Fix Java Docs format

Hackin7 committed:
Update help menu with find command

Hackin7 committed:
Update User Guide with more detail

Hackin7 committed:
Update tests with find command
